### DISCUSSION TITLE
Synchronizing Timer in JMeter DSL

### DISCUSSION THREAD

**Discussion by rsiatka** [abstracta/jmeter-java-dsl#204] [Ideas] [answered]
Hi,

I would like to ask if there are any plans to implement [Synchronizing Timer](https://jmeter.apache.org/usermanual/component_reference.html#Synchronizing_Timer) known from JMeter. I want to send multiple requests simultaneously in batches. I know that JMeter allows to synchronize the execution time of the requests using the mentioned timer.

Here are two cases illustrating the problem and expected behavior:

### _**Sample 1** – 2 threads, 3 iterations, Same user on each iteration, no delay_

![image](https://github.com/abstracta/jmeter-java-dsl/assets/45803776/fc7864db-b96d-4e38-a959-599add85fe0a)

![image](https://github.com/abstracta/jmeter-java-dsl/assets/45803776/8c57e4b4-7bc1-40fa-af92-294fd70a1d99)

![image](https://github.com/abstracta/jmeter-java-dsl/assets/45803776/15cf9699-3c27-4042-9ab6-6aa6064be2aa)

The same effect will be achieved using JMeter DSL - requests are not synchronized (which is perfectly fine when there are no Synchronizing Timer)

```java
    @Test
    public void test1() throws IOException {
        testPlan(
                threadGroup(2, 3,
                        httpSampler("HTTP Request [${__threadNum}-${__counter(TRUE,)}]","https://restcountries.com/v3.1/all")
                        .children(
                                jsr223PostProcessor(s -> s.log.warn(String.valueOf(System.currentTimeMillis()) +
                                        " | " + String.valueOf(s.prev.getConnectTime())))
                        )
                )
        ).run();
    }
```

### _**Sample 2** – 2 threads, 3 iterations, the same user on each iteration, no delay, with Synchronizing Timer_

![image](https://github.com/abstracta/jmeter-java-dsl/assets/45803776/f98d8f3a-be4f-490f-aff3-f0605684e9f6)

![image](https://github.com/abstracta/jmeter-java-dsl/assets/45803776/121b9bee-a677-4fda-b042-f1ade1ef1624)

![image](https://github.com/abstracta/jmeter-java-dsl/assets/45803776/043f6dff-f488-4d35-8d52-20811879ebb4)

![image](https://github.com/abstracta/jmeter-java-dsl/assets/45803776/6d2b9343-d146-4d03-93d4-0a457b9de1da)

As you can see the parallel requests start at the same time (with millisecond accuracy)

**Comment by rabelenda** [comment]
Hello, thank you for bringing this into attention, and providing good context and explanation!

Do you usually just use it with properties set to 0 (number of threads to wait for, and timeout to wait for)? Or do you at some scenarios actually set values for these two properties?

**Comment by rsiatka** [comment]
To be honest I don’t use „Ramp-up period” set to 0 very oftent. Also the „Timeout in miliseconds” from Synchronizing timer is not used by me fequently,  because i always know the numer of executed threads. Nevertheless, this parameter is crucial. 
From what i can see if the NUMBER_OF_THRADES % NUMBER_OF_SIMULATED_USER_GROUPOS = 0 then the „Timeout in miliseconds” is not needed – can be set to 0. But for NUMBER_OF_THRADES % NUMBER_OF_SIMULATED_USER_GROUPOS > 0 the „Timeout in miliseconds” must be set to value bigger than 0 if we don’t want to pause infinitely

**Comment by rabelenda** [comment, marked as answer]
Hello, 

We just released [a new version](https://github.com/abstracta/jmeter-java-dsl/releases/tag/v1.17) which adds `synchronizingTimer` to the DSL. 

[Here](https://abstracta.github.io/jmeter-java-dsl/guide/#requests-synchronization) is the new user guide section. 

Regards

**Reply by rsiatka** [reply]
Quick and perfect. Thank you very much!
